### PR TITLE
Add ShuffleNet V2 to Model ZOO

### PR DESCRIPTION
**Pull request by KsenijaS**

### CODE REVIEW

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+Computationally efficient CNN architecture designed specifically for mobile devices with very limited computing power.
+
+## Description
+[ShuffleNetV2](https://pytorch.org/hub/pytorch_vision_shufflenet_v2/) is an architecture that is the state-of-the-art in terms of speed and accuracy tradeoff.
+
+## Model
+
+|Model        |Download  |Checksum| Download (with sample test data)|ONNX version|Opset version|Top-1 error |Top-5 error |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|ShuffleNetV2 |[5.28MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/model.onnx) | [MD5](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/shufflenetv2-md5.txt) | [5.13MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/model.zip) | 1.6 | 10 | 30.64 | 11.68|
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: Please add models through Git LFS. (i.e. in super-resolution)

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+[ShuffleNetV2](https://pytorch.org/hub/pytorch_vision_shufflenet_v2/) is an architecture that is the state-of-the-art in terms of speed and accuracy tradeoff.
+
+## Model
+
+|Model        |Download  |Checksum| Download (with sample test data)|ONNX version|Opset version|Top-1 error |Top-5 error |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|ShuffleNetV2 |[5.28MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/model.onnx) | [MD5](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/shufflenetv2-md5.txt) | [5.13MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/model.zip) | 1.6 | 10 | 30.64 | 11.68| 
+
+## Inference
+The script for ONNX model conversion and ONNXRuntime inference is [here]
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: Script missing? Also minor nitpick: ONNX Runtime instead of ONNXRuntime

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# ShuffleNetV2
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: All of these files should be moved to vision, not text.

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+## Use-cases
+Computationally efficient CNN architecture designed specifically for mobile devices with very limited computing power.
+
+## Description
+[ShuffleNetV2](https://pytorch.org/hub/pytorch_vision_shufflenet_v2/) is an architecture that is the state-of-the-art in terms of speed and accuracy tradeoff.
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: describe category: object detection? image classification?

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+|Model        |Download  |Checksum| Download (with sample test data)|ONNX version|Opset version|Top-1 error |Top-5 error |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|ShuffleNetV2 |[5.28MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/model.onnx) | [MD5](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/shufflenetv2-md5.txt) | [5.13MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/shufflenet_v2/model/model.zip) | 1.6 | 10 | 30.64 | 11.68| 
+
+## Inference
+The script for ONNX model conversion and ONNXRuntime inference is [here]
+
+### Input to model
+Input to a model are 3-channel RGB images of shape (3 x H x W).
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: nit: shape `(3 x H x W)`
nit: Inputs to the model are 3-channel ...

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+### Preprocessing steps
+All pre-trained models expect input images normalized in the same way, i.e. mini-batches of 3-channel RGB images of shape (3 x H x W), where H and W are expected to be at least 224. The images have to be loaded in to a range of [0, 1] and then normalized using mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225].
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: nit: shape `(3 x H x W)`
224 dimension for H and W should be specified in "Input" section

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+input_tensor = preprocess(input_image)
+```
+Crete a mini-batch as expected by the model
+```python
+input_batch = input_tensor.unsqueeze(0)
+```
+<hr>
+
+## Dataset (Train and Validation)
+Models are pretrainedon ImageNet.
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: nit: pretrained on
add links for image net / coco

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    transforms.ToTensor(),
+    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
+])
+input_tensor = preprocess(input_image)
+```
+Crete a mini-batch as expected by the model
+```python
+input_batch = input_tensor.unsqueeze(0)
+```
+<hr>
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: "Output to model" section?
"Postprocessing steps" section?
i.e. https://github.com/onnx/models/tree/master/vision/style_transfer/fast_neural_style

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
+# ShuffleNetV2
+
+## Use-cases
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: nit: Use cases

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+## Publication/Attribution
+Ningning Ma, Xiangyu Zhang, Hai-Tao Zheng, Jian Sun. ShuffleNet V2: Practical Guidelines for EfficientCNN Architecture Design. 2018.
+<hr>
+
+## Contributors
+Ksenija Stanojevic
+<hr>
+
+## Licence
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: nit: License

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+## Dataset (Train and Validation)
+Models are pretrainedon ImageNet.
+For training we usetrain+valset in COCO except for 5000 images from minivalset, and use the minivalset to test.
+<hr>
+
+## Validation accuracy
+Details of performance on COCO object detection are provided [here](https://arxiv.org/pdf/1807.11164v1.pdf)
+<hr>
+
+## Publication/Attribution
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: Call this section "References"

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+```
+<hr>
+
+## Dataset (Train and Validation)
+Models are pretrainedon ImageNet.
+For training we usetrain+valset in COCO except for 5000 images from minivalset, and use the minivalset to test.
+<hr>
+
+## Validation accuracy
+Details of performance on COCO object detection are provided [here](https://arxiv.org/pdf/1807.11164v1.pdf)
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: Move validation details into Dataset section above and remove this section.

File: text/machine_comprehension/shufflenet_v2/README.md (hunk 4 above)
Comment (vinitra) [reply text/machine_comprehension/shufflenet_v2/README.md]: link paper in description

File: text/machine_comprehension/shufflenet_v2/shufflenet-export.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        input_image = Image.open(filename)
+        preprocess = transforms.Compose([
+            transforms.Resize(256),
+            transforms.CenterCrop(224),
+            transforms.ToTensor(),
+            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
+        ])
+        input_tensor = preprocess(input_image)
+        input_1 = input_tensor.unsqueeze(0)
+        #input_1 = torch.randn(24, 3, 3, 3)
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/shufflenet-export.py]: is this comment needed?

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+## Publication/Attribution
+Ningning Ma, Xiangyu Zhang, Hai-Tao Zheng, Jian Sun. ShuffleNet V2: Practical Guidelines for EfficientCNN Architecture Design. 2018.
+<hr>
+
+## Contributors
+Ksenija Stanojevic
+<hr>
+
+## Licence
+Apache 2.0 License
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: Why Apache 2.0?

File: text/machine_comprehension/shufflenet_v2/shufflenet-export.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    if hasattr(model, 'train'):
+        model.train(False)
+    dir = './'
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+    dir = os.path.join(dir, 'test_' + name)
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+    def f(t):
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/shufflenet-export.py]: clarity in function names or descriptive comments needed

File: text/machine_comprehension/shufflenet_v2/shufflenet-export.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
+
+
+def inference(f, inputs, outputs):
+    sess = onnxruntime.InferenceSession(f)
+    ort_inputs = {sess.get_inputs()[0].name : to_numpy(inputs)}
+    res = sess.run(None, ort_inputs)
+
+    if outputs is not None:
+        print("== CHEKING OUTPUT ==")
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/shufflenet-export.py]: nit: CHECKING

File: text/machine_comprehension/shufflenet_v2/shufflenet-export.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+            transforms.CenterCrop(224),
+            transforms.ToTensor(),
+            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
+        ])
+        input_tensor = preprocess(input_image)
+        input_1 = input_tensor.unsqueeze(0)
+        #input_1 = torch.randn(24, 3, 3, 3)
+        output_1 = model(input_1)
+
+        f = save('shufflenetv2', model.cpu(), input_1, output_1,
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/shufflenet-export.py]: better variable name might be helpful here

File: text/machine_comprehension/shufflenet_v2/README.md [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+## Use-cases
+Computationally efficient CNN architecture designed specifically for mobile devices with very limited computing power.
+
+## Description
+[ShuffleNetV2](https://pytorch.org/hub/pytorch_vision_shufflenet_v2/) is an architecture that is the state-of-the-art in terms of speed and accuracy tradeoff.
+
+## Model
+
+|Model        |Download  |Checksum| Download (with sample test data)|ONNX version|Opset version|Top-1 error |Top-5 error |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
```
Comment (vinitra) [inline text/machine_comprehension/shufflenet_v2/README.md]: You can put the Checksum directly in the table like: https://github.com/onnx/models/tree/master/vision/style_transfer/fast_neural_style

File: vision/classification/shufflenet_v2/README.md [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+```
+
+### Output of model
+
+Output of this model is tensor of shape 1000, with confidence scores over Imagenet's 1000 classes.
+
+<hr>
+
+## Dataset (Train and Validation)
+Models are pretrainedon ImageNet.
```
Comment (prasanthpul) [inline vision/classification/shufflenet_v2/README.md]: missing space "pretrained on"

File: vision/classification/shufflenet_v2/README.md [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+For training we usetrain+valset in COCO except for 5000 images from minivalset, and use the minivalset to test.
```
Comment (prasanthpul) [inline vision/classification/shufflenet_v2/README.md]: missing space "use train"

File: vision/classification/shufflenet_v2/shufflenet-export.py [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+#           GitHub Repo    |    Model   
+MODELS = [
+    ('pytorch/vision:v0.5.0', 'shufflenet_v2_x0_5'),
+    ('pytorch/vision:v0.5.0', 'shufflenet_v2_x1_0'),
+]
+data_dir = 'test_data_set_0'
+
+url, filename = ("https://github.com/pytorch/hub/raw/master/dog.jpg", "dog.jpg")
+try: urllib.URLopener().retrieve(url, filename)
+except: urllib.request.urlretrieve(url, filename)
```
Comment (prasanthpul) [inline vision/classification/shufflenet_v2/shufflenet-export.py]: please try to address the issue identified by LGTM

File: vision/classification/shufflenet_v2/shufflenet-export.py [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+    if hasattr(model, 'train'):
+        model.train(False)
+    dir = './'
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+    dir = os.path.join(dir, 'test_' + name)
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+    def f(t):
```
Comment (vinitra) [inline vision/classification/shufflenet_v2/shufflenet-export.py]: I still feel that names f and g could be more specific to clarify for new users -- unless you have a reason for naming the helper functions this way?

**Comment by KsenijaS** [comment]
@prasanthpul @EmmaNingMS Vinitra approved this PR. Can it be merged now?

**Comment by prasanthpul** [comment]
do you mind resolving all the code review conversations that you addressed? they are all still showing as open, so hard to verify everything is resolved (I assume they are)

**Comment by KsenijaS** [comment]
@prasanthpul I marked everything as resolved now
